### DISCUSSION TITLE
Restoring workout record

### DISCUSSION THREAD

**Discussion by TClin76** [dvmarinoff/Auuki#94] [General]
Hi @dvmarinoff 

If Chrome crash before workout ending, is it still possible to find the temporary record (cache?) in any location for restoring .fit file?
Thanks

**Comment by dvmarinoff** [comment]
Data of only the current running workout is stored into the browser IndexedDB (that's a form of browser cache). It gets discarded only when the user clicks the stop button and reloads the page. For as long as you don't end up (stop) the workout it should survive page reloads, closing the tab or the whole browser. I have it once keep data for me over 3 days. The side effect of this is that if you continue your workout 3 days later it will look as a 4 days long workout, so you'll need to manually edit the timestamps with something like [fitfiletools](https://www.fitfiletools.com/#/top).

When you open up the app again if you see numbers in the interval and elapsed timer slots the data is still there, you can continue from this point or stop and save, if you see dashes the data backup failed

Now browser crash might be a different case. Does it appear to be caused by the app, or something else?

I don't won't to introduce a remote data base, and for a year of using the app I haven’t lost any data,
but I'll think for other ways make data storage more robust.

Did a test today:
https://youtu.be/uwFc-KNKoWY

**Comment by TClin76** [comment]
I got Chrome crash twice when the workout is about to finish and I want to take a screenshot (Shift+Command+3). 

I remembered when I re-start the browser, the timer didn’t continue. Probably crash is not the same as closing the browser?

**Reply by dvmarinoff** [reply]
I have never seen a crash with Chrome, maybe ones it force closed and opened again after an update. Do you have any ideas on what might be the cause of the crash (chrome plugins, other apps, etc.) ?

**Reply by dvmarinoff** [reply]
Do you had other crashes before or is it something recent? Do you mean that Shift+Cmd+3 is crashing it?

**Reply by TClin76** [reply]
Yes. Both crashes were happened when I press Shift+Cmd+3 to capture the screen. But not every Shift+cmd+3 cause crash.

**Comment by dvmarinoff** [comment]
It's not cool to loose data, I am working on this, hope to deploy a fix today, but until then one way to force data backup is to switch to another open tab in the browser, and then go back to the app.

I have the app listen for reload or release (tab close or switch or just browser going in the background on mobile devices) event and when that happens it backs up the data, my guess is that this doesn't get processed when the browser just crashes. So I am going to add code that makes a backup every 60 seconds, and I'll try to do some load testing.

**Reply by TClin76** [reply]
I think so. It seems not able to continue when Browser is not closed in correct way.
Backup in every 60 seconds is nice, just like what Microsoft office does for Autosave.

**Comment by dvmarinoff** [comment]
Deployed a solution that is backing up data every 60 seconds.

Did a crash test with an old thinkpad with Ubuntu on it:
- cycled for 1:30
- pressed the turn off button until the laptop force shut downs
- turned on the machine, launched chrome and opened the app (not using the restore option)
- the timer was at 60s with 60 records of data available in the browser data base

1. original session
![Screenshot from 2022-01-16 15-45-57](https://user-images.githubusercontent.com/7523832/149663152-7ec68699-f376-4562-81d5-57a00beb8382.jpg)

2. launching chrome after the forced shut down
![Screenshot from 2022-01-16 15-46-51](https://user-images.githubusercontent.com/7523832/149663161-702411b0-5123-4e8d-9434-b4f67b97625d.jpg)

3. recovered session
![Screenshot from 2022-01-16 15-48-01](https://user-images.githubusercontent.com/7523832/149663246-3e30603b-8060-4946-8ea4-6d30402226b1.jpg)

**Reply by TClin76** [reply]
Do you mean not using its pop-up restore button, but open Flux manually?

**Comment by TClin76** [comment]
I check the folder of IndexedDB, could you identify which file is the missing data?
<img width="388" alt="截圖 2022-01-16 下午10 31 29" src="https://user-images.githubusercontent.com/96434118/149664252-28fc1f2c-ea4c-4780-b6c6-4d4acee7c4fa.png">

**Reply by dvmarinoff** [reply]
No idea, but definitely make a copy of those folders. I'll look into it.

**Reply by TClin76** [reply]
According to the file size and modified time, below file is more likely to be the data recorded during workout.
[2.zip](https://github.com/dvmarinoff/Flux/files/7879836/2.zip)
